crates/sui-display/src/v1/lexer.rs

```diff
@@ -0,0 +1,370 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt;
+
+/// Lexer for Display V1 format strings. Format strings are a mix of text and expressions.
+/// Expressions are enclosed in curly braces and may contain identifiers separated by dots (a path
+/// of field accesses).
+#[derive(Debug)]
+pub(crate) struct Lexer<'s> {
+    /// Remaining input to be tokenized.
+    src: &'s str,
+
+    /// The number of bytes tokenized so far.
+    off: usize,
+
+    /// Whether the lexer is currently inside a text strand or an expression strand.
+    mode: Mode,
+}
+
+#[derive(Debug)]
+enum Mode {
+    Text,
+    Expr,
+}
+
+/// A lexeme is a token along with its offset in the source string, and the slice of source string
+/// that it originated from.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(crate) struct Lexeme<'s>(pub Token, pub usize, pub &'s str);
+
+/// Like [Lexeme] but owns the slice of source string. Useful for capturing context in an error
+/// message.
+#[derive(Debug)]
+pub(crate) struct OwnedLexeme(Token, usize, String);
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(crate) enum Token {
+    /// '.'
+    Dot,
+    /// '\X' where X is any byte.
+    Escaped,
+    /// A potential field identifier
+    Ident,
+    /// '{'
+    LCurl,
+    /// '}'
+    RCurl,
+    /// A strand of text.
+    Text,
+}
+
+#[derive(Debug)]
+pub(crate) struct TokenSet<'t>(pub &'t [Token]);
+
+impl<'s> Lexer<'s> {
+    pub(crate) fn new(src: &'s str) -> Self {
+        Self {
+            src,
+            off: 0,
+            mode: Mode::Text,
+        }
+    }
+
+    /// Assuming the lexer is in text mode, return the next text token.
+    fn next_text_token(&mut self) -> Option<Lexeme<'s>> {
+        let bytes = self.src.as_bytes();
+
+        use Token as T;
+        Some(match bytes.first()? {
+            b'\\' if bytes.len() > 1 => {
+                self.take(T::Escaped, 1); // discard the backslash
+                self.take(T::Escaped, 1)
+            }
+            b'\\' => self.take(T::Text, 1),
```

Comment:
Looks like it should be fine, but if you want to write more tests... testing weird escapes like `\\\\\\\\\\\` or similar may be nice to have. Not necessary to land though.

Reply:
Done, thanks for the call out!